File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3022_minimize-or-of-remaining-elements-using-operations/solutions/cpp/solution.cpp
#include <vector>
#include <queue>
#include <functional>

class Solution {
public:
    int minOrAfterOperations(std::vector<int>& nums, int k) {
        int n = nums.size();
        if (k == 0) {
            int result = 0;
            for (int num : nums) {
                result |= num;
            }
            return result;
        }
        
        // Priority queue to store the pairs (OR value, length of segment)
        std::priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>, std::greater<>> pq;
        
        // Initially, each element is its own segment
        for (int num : nums) {
            pq.push({num, 1});
        }
        
        while (k > 0 && pq.size() > 1) {
            auto first = pq.top(); pq.pop();
            auto second = pq.top(); pq.pop();
            
            // Combine the two smallest segments
            int combined = first.first & second.first;
            pq.push({combined, first.second + second.second});
            k--;
        }
        
        // Compute the final OR of remaining elements in the queue
        int finalOr = 0;
        while (!pq.empty()) {
            finalOr |= pq.top().first;
            pq.pop();
        }
        
        return finalOr;
    }
};

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3022_minimize-or-of-remaining-elements-using-operations/solutions/java/solution.java
class Solution {
    public int minOrAfterOperations(int[] nums, int k) {
        int n = nums.length;
        if (k == 0) {
            int res = 0;
            for (int num : nums) {
                res |= num;
            }
            return res;
        }

        // Priority queue to store pairs of (bitwise OR of segment, size of segment)
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[1]));

        // Initialize with each element as a segment
        for (int num : nums) {
            pq.offer(new int[] { num, 1 });
        }

        // Perform up to k operations
        for (int i = 0; i < k; i++) {
            if (pq.size() < 2) break;

            // Take two smallest segments
            int[] first = pq.poll();
            int[] second = pq.poll();

            // Combine them
            int combinedOr = first[0] | second[0];
            int newSize = first[1] + second[1];
            pq.offer(new int[] { combinedOr, newSize });
        }

        // Calculate the final OR of remaining segments
        int finalOr = 0;
        while (!pq.isEmpty()) {
            finalOr |= pq.poll()[0];
        }

        return finalOr;
    }
}
